Helidon Inject Examples

romain-grecourt opened the pull request:
Add examples for Helidon Inject.

Fixes #112

romain-grecourt committed:
Helidon Inject Examples

tomas-langer commented on examples/inject/src/main/java/io/helidon/examples/inject/CovarianceExample.java:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    public static void main(String[] args) {
+        var registry = ServiceRegistryManager.create().registry();
```
The examples should either use the `Services` class, or create a manager and shut it down when finished.
When using method `create()`, the registry manager is not registered with a shutdown hook, so `@PreDestroy` methods would not get called.

romain-grecourt replied:
Should it implement AutoCloseable ?

romain-grecourt replied:
I have replaced most usages with Services.
There are two classes that still use ServiceRegistryManager, I've used a try/finally block in those.

tomas-langer replied:
I would prefer this fixed, as this will get re-arranged when I run code format, and that is the reason why we have code format specified in the project...

tomas-langer commented on examples/inject/src/main/java/io/helidon/examples/inject/CovarianceExample.java:
```diff
@@ -0,0 +1,76 @@
+    record Shelter(List<Pet> all, List<Cat> cats, List<Dog> dogs) {
+    }
+
+    public static void main(String[] args) {
```
In all our classes (and in our code formatting setup) we have methods first, classes later.

romain-grecourt replied:
It messes the reading, because the main class code comes before.
Can we make an exception or shall I go ahead and move the main methods in all the files ?

tomas-langer replied:
I would prefer this fixed, as this will get re-arranged when I run code format, and that is the reason why we have code format specified in the project...

romain-grecourt replied:
Specified where ? The checkstyle check is passing.

romain-grecourt replied:
I have moved all the method methods at the top.

tomas-langer replied:
Specified here: https://github.com/helidon-io/helidon/blob/main/etc/codestyle/idea-code-style.xml

romain-grecourt replied:
Unless we can run this file in a GitHub action, we shouldn't require any changes based on it since we can't enforce anything.

tomas-langer commented on examples/inject/src/main/java/io/helidon/examples/inject/FactoryExample.java:
```diff
@@ -0,0 +1,157 @@
+    /**
+     * A bulk factory that provides qualified {@link Color} instances.
+     */
+    @Red
```
This is no longer needed. A ServicesFactory is considered a "multi-qualifier" factory by default, so it should work without these annotations.

romain-grecourt replied:
It does not work without the annotations:

```
Exception in thread "main" io.helidon.service.registry.ServiceRegistryException: Neither of matching services could provide a value. Descriptors: [], lookup: Lookup{serviceType=Optional.empty,scopes=[],qualifiers=[Qualifier{};Annotation{typeName=io.helidon.examples.inject.FactoryExample.Red,values={}}],contracts=[io.helidon.examples.inject.FactoryExample.Color],contractType=Optional[class io.helidon.examples.inject.FactoryExample$Color],runLevel=Optional.empty,weight=Optional.empty,includeAbstract=false,dependency=Optional[Dependency{service=io.helidon.examples.inject.FactoryExample.Colors,name=red,typeName=io.helidon.examples.inject.FactoryExample.Color,elementKind=CONSTRUCTOR,qualifiers=[Qualifier{};Annotation{typeName=io.helidon.examples.inject.FactoryExample.Red,values={}}],method=Optional.empty,cardinality=REQUIRED,isServiceInstance=false,isSupplier=false}],factoryTypes=[]}
	at io.helidon.service.registry.ServiceSupplies$ServiceSupply.lambda$new$0(ServiceSupplies.java:161)
	at java.base/java.util.Optional.orElseThrow(Optional.java:403)
	at io.helidon.service.registry.ServiceSupplies$ServiceSupply.lambda$new$1(ServiceSupplies.java:161)
	at io.helidon.service.registry.ServiceSupplies$ServiceSupply.get(ServiceSupplies.java:170)
	at io.helidon.service.registry.BindingDependencyContext.dependency(BindingDependencyContext.java:34)
	at io.helidon.examples.inject.FactoryExample_Colors__ServiceDescriptor.instantiate(FactoryExample_Colors__ServiceDescriptor.java:127)
	at io.helidon.examples.inject.FactoryExample_Colors__ServiceDescriptor.instantiate(FactoryExample_Colors__ServiceDescriptor.java:38)
	at io.helidon.service.registry.Activators$InstanceHolderImpl.construct(Activators.java:944)
	at io.helidon.service.registry.Activators$SingleServiceActivator.construct(Activators.java:498)
	at io.helidon.service.registry.Activators$BaseActivator.doActivate(Activators.java:369)
	at io.helidon.service.registry.Activators$BaseActivator.activate(Activators.java:233)
	at io.helidon.service.registry.Activators$BaseActivator.instances(Activators.java:188)
	at io.helidon.service.registry.ServiceSupplies.oneInstance(ServiceSupplies.java:43)
	at io.helidon.service.registry.ServiceSupplies$ServiceSupply.lambda$new$1(ServiceSupplies.java:159)
	at io.helidon.service.registry.ServiceSupplies$ServiceSupply.get(ServiceSupplies.java:170)
	at io.helidon.service.registry.CoreServiceRegistry.get(CoreServiceRegistry.java:180)
	at io.helidon.service.registry.CoreServiceRegistry.get(CoreServiceRegistry.java:175)
	at io.helidon.service.registry.ServiceRegistry.get(ServiceRegistry.java:48)
	at io.helidon.examples.inject.FactoryExample.main(FactoryExample.java:150)
```

tomas-langer replied:
Is it tested in the test? I have removed all three qualifiers from the `io.helidon.examples.inject.FactoryExample.MyBulkServiceFactory` and the build passed.

romain-grecourt replied:
It passes for me today `\_(ツ)_/` ; I was likely working off an outdated workspace.

tomas-langer commented on examples/inject/src/main/java/io/helidon/examples/inject/ExternalContractExample.java:
```diff
@@ -0,0 +1,73 @@
+     * A service that implements {@link CharSequence} contract.
+     */
+    @Service.PerLookup
+    @Service.ExternalContracts(CharSequence.class)
```
`CharSequence` is an interface and as such will be an automatic contract. External contract is needed for classes only (i.e. an abstract class)

romain-grecourt replied:
Roger, used an abstract class instead.

tomas-langer commented on examples/inject/src/main/java/io/helidon/examples/inject/InjectionPointsExample.java:
```diff
@@ -0,0 +1,288 @@
+     * @param optionalGreeter optional greeter
+     */
+    @Service.Singleton
+    record GreetingWithOptionalIP(@Service.Named("non-existing") Optional<Greeter> optionalGreeter) implements Greeting {
```
Please use `Ip` instead of `IP`

romain-grecourt committed:
Review feedback
- Use Services instead of ServiceRegistryManager for the simple usecases
- Use a try-finally shutdown for the usecase that require using ServiceRegistryManager
- Use an abstract class for the external contract example
- Rename GreetingWithOptionalIP to GreetingWithOptionalIp
- Add helidon-logging-jul

romain-grecourt committed:
Add a preview disclaimer

Verdent commented on examples/inject/src/main/java/io/helidon/examples/inject/RunLevelExample.java:
```diff
@@ -0,0 +1,104 @@
+     *
+     * @param registry registry
+     */
+    static void startRunLevels(ServiceRegistry registry) {
```
I think for better understanding of how RunLevels work and what they are used for, would be better to use `start` method on the `ServiceRegistryManager` instead of making here these startRunLevels methods. It makes it more confusing, from my point of view. 

Then all you need to do is the following:
```java
public static void main(String[] args) {
    ServiceRegistryManager registryManager = ServiceRegistryManager.start();
    STARTUP_EVENTS.forEach(System.out::println);
    registryManager.shutdown();
}
```

romain-grecourt replied:
The start method was added after the initial writing of the examples. Makes sense to use it.

romain-grecourt committed:
Review feedback:
- Move main methods at the top
- Remove qualifiers from the ServicesFactory example
- Use ServiceRegistryManager.start for the runLevel example

Other changes:
- Avoid Services in unit test to avoid conflicts between tests
